File: node/src/http.ts
/**
 * HTTP-транспорт: заголовки, ретраи, таймаут, разбор ошибок.
 *
 * Зависимостей нет: используется глобальный `fetch` (в Node 20+ это
 * реализация на базе undici, включённая в рантайм). Для тестов на
 * моках функция fetch внедряется через опцию `fetch` конструктора
 * `WataClient` — это даёт более чистые и быстрые тесты, чем поднятие
 * `undici.MockAgent`, и не требует дополнительной зависимости.
 */

import {
  WataApiError,
  WataAuthError,
  WataNetworkError,
  WataRateLimitError,
  WataServerError,
  WataError,
} from './errors.js';

export type WataFetch = typeof fetch;

export interface TransportOptions {
  baseUrl: string;
  token?: string;
  timeoutMs: number;
  maxRetries: number;
  fetchImpl: WataFetch;
}

export type HttpMethod = 'GET' | 'POST';

export interface RequestOptions {
  method: HttpMethod;
  path: string;
  query?: Record<string, unknown>;
  body?: unknown;
  /** `false` — не добавлять заголовок Authorization (например, публичный ключ вебхуков). */
  auth?: boolean;
  /**
   * Можно ли повторять запрос при сетевой ошибке/5xx. По умолчанию —
   * только для GET. Изменяющие запросы (создание ссылки/платежа/
   * возврата) по умолчанию НЕ повторяются: повтор может создать
   * второй платёж.
   */
  idempotent?: boolean;
}

interface WataErrorBody {
  error?: {
    code?: string;
    message?: string;
    details?: unknown;
    validationErrors?: unknown;
  };
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** Убирает `undefined`/`null` из query-параметров и сериализует массивы как повторяющиеся ключи. */
export function buildQueryString(query?: Record<string, unknown>): string {
  if (!query) return '';
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(query)) {
    if (value === undefined || value === null) continue;
    if (Array.isArray(value)) {
      for (const item of value) {
        if (item === undefined || item === null) continue;
        params.append(key, String(item));
      }
    } else if (value instanceof Date) {
      params.append(key, value.toISOString());
    } else {
      params.append(key, String(value));
    }
  }
  const serialized = params.toString();
  return serialized ? `?${serialized}` : '';
}

/** Рекурсивно убирает `undefined`/`null` поля из тела запроса перед сериализацией в JSON. */
function pruneNullish(value: unknown): unknown {
  if (value === null || value === undefined) return undefined;
  if (Array.isArray(value)) return value.map(pruneNullish);
  if (typeof value === 'object') {
    if (value instanceof Date) return value.toISOString();
    const out: Record<string, unknown> = {};
    for (const [key, item] of Object.entries(value as Record<string, unknown>)) {
      const pruned = pruneNullish(item);
      if (pruned !== undefined) out[key] = pruned;
    }
    return out;
  }
  return value;
}

/**
 * Строит query-объект с ключами в PascalCase, ожидаемом сервером WATA,
 * из объекта с камелкейс-полями SDK. Отсутствующие (undefined/null)
 * значения не попадают в результат.
 */
export function mapQuery<P extends object>(params: P, keyMap: Record<string, string>): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(params as Record<string, unknown>)) {
    if (value === undefined || value === null) continue;
    out[keyMap[key] ?? key] = value;
  }
  return out;
}

export class Transport {
  constructor(private readonly options: TransportOptions) {}

  /**
   * Как `request()`, но добавляет к разобранному ответу поле `raw` с
   * исходным телом целиком. Используется моделями цифровых товаров:
   * документация WATA покрывает не все возвращаемые поля, и без
   * сохранения исходного JSON интегратор не смог бы добраться до
   * недокументированного значения, не дожидаясь новой версии SDK.
   *
   * Ожидает, что сервер отвечает JSON-объектом (не массивом/примитивом);
   * для эндпоинтов, возвращающих массив, оборачивайте результат `request()` вручную.
   */
  async requestWithRaw<Res extends { raw: unknown }>(request: RequestOptions): Promise<Res> {
    const parsed = (await this.request<Record<string, unknown>>(request)) ?? {};
    return { ...parsed, raw: parsed } as Res;
  }

  async request<Res>(request: RequestOptions): Promise<Res> {
    const useAuth = request.auth !== false;
    const url = this.options.baseUrl + request.path + buildQueryString(request.query);
    const allowRetry = request.idempotent ?? request.method === 'GET';
    const maxAttempts = allowRetry ? Math.max(1, this.options.maxRetries) : 1;

    let attempt = 0;
    for (;;) {
      attempt++;
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), this.options.timeoutMs);
      try {
        const headers: Record<string, string> = {
          Accept: 'application/json',
        };
        if (request.body !== undefined) headers['Content-Type'] = 'application/json';
        if (useAuth) headers.Authorization = `Bearer ${this.options.token}`;

        const init: RequestInit = {
          method: request.method,
          headers,
          signal: controller.signal,
        };
        if (request.body !== undefined) {
          init.body = JSON.stringify(pruneNullish(request.body));
        }

        const response = await this.options.fetchImpl(url, init);

        if (response.status === 429) {
          throw await this.toRateLimitError(response, request.path);
        }
        if (response.status === 401 || response.status === 403) {
          throw await this.toAuthError(response, request.path);
        }
        if (response.status >= 500) {
          const serverError = await this.toServerError(response, request.path);
          if (attempt < maxAttempts) {
            await this.backoff(attempt);
            continue;
          }
          throw serverError;
        }
        if (response.status >= 400) {
          throw await this.toApiError(response, request.path);
        }

        return await this.parseBody<Res>(response);
      } catch (err) {
        if (err instanceof WataError) throw err;
        const networkError = new WataNetworkError(this.describeNetworkError(err), {
          path: request.path,
          cause: err,
        });
        if (allowRetry && attempt < maxAttempts) {
          await this.backoff(attempt);
          continue;
        }
        throw networkError;
      } finally {
        clearTimeout(timer);
      }
    }
  }

  private describeNetworkError(err: unknown): string {
    if (err instanceof Error && err.name === 'AbortError') {
      return `Превышен таймаут ожидания ответа (${this.options.timeoutMs} мс).`;
    }
    return `Сетевая ошибка при обращении к WATA API: ${err instanceof Error ? err.message : String(err)}`;
  }

  private async backoff(attempt: number): Promise<void> {
    const base = 200 * 2 ** (attempt - 1);
    const jitter = Math.random() * base * 0.5;
    await sleep(base + jitter);
  }

  private async parseBody<Res>(response: Response): Promise<Res> {
    if (response.status === 204) return undefined as Res;
    const text = await response.text();
    if (!text) return undefined as Res;
    return JSON.parse(text) as Res;
  }

  private async readErrorBody(response: Response): Promise<WataErrorBody | undefined> {
    try {
      const text = await response.text();
      return text ? (JSON.parse(text) as WataErrorBody) : undefined;
    } catch {
      return undefined;
    }
  }

  private async toApiError(response: Response, path: string): Promise<WataApiError> {
    const body = await this.readErrorBody(response);
    const message = body?.error?.message ?? `WATA API вернул ошибку ${response.status} для ${path}.`;
    return new WataApiError(message, {
      httpStatus: response.status,
      path,
      wataCode: body?.error?.code,
      details: body?.error?.details,
      validationErrors: body?.error?.validationErrors,
    });
  }

  private async toAuthError(response: Response, path: string): Promise<WataAuthError> {
    const body = await this.readErrorBody(response);
    const message =
      body?.error?.message ??
      'Ошибка авторизации: токен недействителен, отозван, принадлежит другому терминалу либо запрос выполнен с несогласованного IP.';
    return new WataAuthError(message, { httpStatus: response.status, path, wataCode: body?.error?.code });
  }

  private async toServerError(response: Response, path: string): Promise<WataServerError> {
    return new WataServerError(`WATA API вернул серверную ошибку ${response.status} для ${path}.`, {
      httpStatus: response.status,
      path,
    });
  }

  private async toRateLimitError(response: Response, path: string): Promise<WataRateLimitError> {
    const retryAfterHeader = response.headers.get('Retry-After');
    let retryAfterMs: number | undefined;
    let retryAt: Date | undefined;
    if (retryAfterHeader) {
      const seconds = Number(retryAfterHeader);
      if (Number.isFinite(seconds)) {
        retryAfterMs = seconds * 1000;
        retryAt = new Date(Date.now() + retryAfterMs);
      } else {
        const date = new Date(retryAfterHeader);
        if (!Number.isNaN(date.getTime())) retryAt = date;
      }
    }
    const suffix = retryAt ? ` Повторите после ${retryAt.toISOString()}.` : '';
    return new WataRateLimitError(`Превышен лимит запросов (429) для ${path}.${suffix}`, {
      httpStatus: 429,
      path,
      retryAfterMs,
      retryAt,
    });
  }
}

File: go/http.go
package wata

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"math/rand"
	"net/http"
	"net/url"
	"strconv"
	"time"
)

// apiRequest describes one HTTP call to a WATA product API.
type apiRequest struct {
	Method string
	// Path is appended to the product's path prefix ("/api/h2h" for
	// acquiring, "/api" for digital goods), e.g. "/links".
	Path  string
	Query url.Values
	// Body, if non-nil, becomes the JSON request body. It may be a Go value
	// (marshaled with encoding/json) or pre-encoded []byte / json.RawMessage
	// (used as-is) — the latter is how services merge a typed, documented
	// field with caller-supplied extra fields for endpoints SPEC.md does not
	// fully enumerate.
	Body any
	// Mutating marks a state-changing request (create link/payment/refund).
	// SPEC.md requires these to never be retried automatically, since a
	// retry could create a second payment. Mutating requests are attempted
	// exactly once regardless of WithMaxRetries.
	Mutating bool
	// NoAuth is set only for GET /api/h2h/public-key, the sole endpoint
	// called without an Authorization header.
	NoAuth bool
}

const userAgent = "wata-sdk-go/" + SDKVersion

func (c *Client) do(ctx context.Context, p product, r apiRequest, out any) error {
	var token string
	if !r.NoAuth {
		t, err := c.tokenFor(p)
		if err != nil {
			return err
		}
		token = t
	}

	base, err := c.baseURL(p)
	if err != nil {
		return err
	}

	fullPath := p.pathPrefix() + r.Path
	u, err := url.Parse(base + fullPath)
	if err != nil {
		return newConfigError(fmt.Sprintf("некорректный путь запроса %q: %v", fullPath, err))
	}
	if len(r.Query) > 0 {
		u.RawQuery = r.Query.Encode()
	}

	var bodyBytes []byte
	switch b := r.Body.(type) {
	case nil:
	case []byte:
		bodyBytes = b
	case json.RawMessage:
		bodyBytes = b
	default:
		data, err := json.Marshal(b)
		if err != nil {
			return newConfigError(fmt.Sprintf("не удалось сериализовать тело запроса: %v", err))
		}
		bodyBytes = data
	}

	// Mutating requests (create link/payment/refund) are never retried
	// automatically — a retry after a network blip could create a second
	// payment. GET requests, including the unauthenticated public-key
	// fetch, retry on network errors and 5xx per SPEC.md section 3.
	retryable := !r.Mutating

	attempts := 1
	if retryable {
		attempts = c.maxRetries
		if attempts < 1 {
			attempts = 1
		}
	}

	var lastErr error
	for attempt := 0; attempt < attempts; attempt++ {
		if attempt > 0 {
			if werr := waitBackoff(ctx, attempt); werr != nil {
				return lastErr
			}
		}

		reqCtx, cancel := context.WithTimeout(ctx, c.timeout)
		var bodyReader io.Reader
		if bodyBytes != nil {
			bodyReader = bytes.NewReader(bodyBytes)
		}
		httpReq, err := http.NewRequestWithContext(reqCtx, r.Method, u.String(), bodyReader)
		if err != nil {
			cancel()
			return &NetworkError{Err: &Error{Message: err.Error(), Path: fullPath}, Cause: err}
		}
		if bodyBytes != nil {
			httpReq.Header.Set("Content-Type", "application/json")
		}
		httpReq.Header.Set("Accept", "application/json")
		httpReq.Header.Set("User-Agent", userAgent)
		if !r.NoAuth {
			httpReq.Header.Set("Authorization", "Bearer "+token)
		}

		resp, err := c.httpClient.Do(httpReq)
		if err != nil {
			cancel()
			netErr := &NetworkError{Err: &Error{Message: err.Error(), Path: fullPath}, Cause: err}
			lastErr = netErr
			if retryable && attempt < attempts-1 {
				continue
			}
			return netErr
		}

		data, readErr := io.ReadAll(resp.Body)
		closeErr := resp.Body.Close()
		cancel()
		if readErr != nil {
			netErr := &NetworkError{
				Err:   &Error{Message: readErr.Error(), HTTPStatus: resp.StatusCode, Path: fullPath},
				Cause: readErr,
			}
			lastErr = netErr
			if retryable && attempt < attempts-1 {
				continue
			}
			return netErr
		}
		if closeErr != nil {
			// Body was fully read already; a close error here is not fatal
			// to interpreting the response we already have.
			_ = closeErr
		}

		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
			if out != nil && len(data) > 0 {
				if err := json.Unmarshal(data, out); err != nil {
					return newConfigError(fmt.Sprintf("не удалось разобрать ответ WATA (%s %s): %v", r.Method, fullPath, err))
				}
				if rc, ok := out.(rawCapturer); ok {
					rc.captureRawJSON(data)
				}
			}
			return nil
		}

		if resp.StatusCode >= 500 {
			srvErr := &ServerError{Err: &Error{
				Message:    serverErrorMessage(data),
				HTTPStatus: resp.StatusCode,
				Path:       fullPath,
			}}
			lastErr = srvErr
			if retryable && attempt < attempts-1 {
				continue
			}
			return srvErr
		}

		// Non-retryable 4xx: return immediately, do not loop.
		return errorFromResponse(resp.StatusCode, fullPath, data, resp.Header)
	}
	return lastErr
}

// waitBackoff sleeps with exponential backoff and jitter before retry
// attempt number `attempt` (1-based: called before the 2nd, 3rd, ... try).
// It returns ctx.Err() if ctx is done before the wait completes.
func waitBackoff(ctx context.Context, attempt int) error {
	base := 200 * time.Millisecond
	backoff := time.Duration(float64(base) * math.Pow(2, float64(attempt-1)))
	const maxBackoff = 5 * time.Second
	if backoff > maxBackoff {
		backoff = maxBackoff
	}
	jitter := time.Duration(rand.Int63n(int64(backoff/2) + 1))
	wait := backoff/2 + jitter

	t := time.NewTimer(wait)
	defer t.Stop()
	select {
	case <-t.C:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

// apiErrorBody mirrors the {error: {code, message, details,
// validationErrors}} shape SPEC.md documents for 4xx responses.
type apiErrorBody struct {
	Error struct {
		Code             string            `json:"code"`
		Message          string            `json:"message"`
		Details          any               `json:"details"`
		ValidationErrors []ValidationError `json:"validationErrors"`
	} `json:"error"`
}

func errorFromResponse(status int, path string, data []byte, header http.Header) error {
	var body apiErrorBody
	_ = json.Unmarshal(data, &body) // best-effort: some 4xx bodies may not match the shape.

	msg := body.Error.Message
	if msg == "" {
		msg = http.StatusText(status)
	}
	code := body.Error.Code

	switch status {
	case http.StatusUnauthorized, http.StatusForbidden:
		return &AuthError{Err: &Error{Message: msg, HTTPStatus: status, Path: path, Code: code}}
	case http.StatusTooManyRequests:
		return &RateLimitError{
			Err:        &Error{Message: msg, HTTPStatus: status, Path: path, Code: code},
			RetryAfter: parseRetryAfter(header),
		}
	default:
		return &APIError{
			Err:              &Error{Message: msg, HTTPStatus: status, Path: path, Code: code},
			Details:          body.Error.Details,
			ValidationErrors: body.Error.ValidationErrors,
		}
	}
}

func parseRetryAfter(header http.Header) *time.Duration {
	v := header.Get("Retry-After")
	if v == "" {
		return nil
	}
	if secs, err := strconv.Atoi(v); err == nil {
		d := time.Duration(secs) * time.Second
		return &d
	}
	if t, err := http.ParseTime(v); err == nil {
		d := time.Until(t)
		return &d
	}
	return nil
}

func serverErrorMessage(data []byte) string {
	var body apiErrorBody
	if err := json.Unmarshal(data, &body); err == nil && body.Error.Message != "" {
		return body.Error.Message
	}
	if len(data) > 0 && len(data) < 500 {
		return string(data)
	}
	return "внутренняя ошибка сервера WATA"
}

// marshalWithExtra marshals known to JSON, then overlays the keys of extra
// on top of the resulting object (extra wins on key conflicts). It is used
// by digital-goods request builders to combine a typed, spec-documented
// field (e.g. netAmount) with additional fields the live API accepts but
// SPEC.md does not name.
func marshalWithExtra(known any, extra map[string]any) (json.RawMessage, error) {
	base, err := json.Marshal(known)
	if err != nil {
		return nil, err
	}
	if len(extra) == 0 {
		return base, nil
	}
	m := map[string]any{}
	if err := json.Unmarshal(base, &m); err != nil {
		return nil, err
	}
	for k, v := range extra {
		m[k] = v
	}
	return json.Marshal(m)
}
